### DISCUSSION TITLE
Does fluxcd always push custom resources to version control before applying?

### DISCUSSION THREAD

**Discussion by RytisLT** [fluxcd/flux2#2580] [Q&A] [answered]
Hello, I'm following `Flux for Helm Users` guide and failing at first step:

    $ flux create source helm traefik --url https://helm.traefik.io/traefik --namespace traefik

the command fails with

    ✗ namespaces "traefik" not found

I understand that the problem is the missing namespace but the guide never mentions it. I don't know whether I should create it manually with `kubectl` or using flux or maybe flux should create the missing namespace. 

If I use default namespace the command completes successfully

    $ flux create source helm traefik --url https://helm.traefik.io/traefik
    ✚ generating HelmRepository source
    ► applying HelmRepository source
    ✔ source created
    ◎ waiting for HelmRepository source reconciliation
    ✔ HelmRepository source reconciliation completed
    ✔ fetched revision: c5990383f33754f7465448b9e7198cf06c6fed645ad48b7fff595effe74b5826

Yet there is no change to the source code in my repository. I though that flux will create a Custom Resource push it to the repo and then apply it from there. Is my understanding how flux works totally wrong or maybe there's something wrong with my flux installation (I've bootstrapped with gitlab PAT)?

**Comment by kingdonb** [comment, marked as answer]
Thanks for posting, we don't mean for anyone to get stuck! So it's very helpful to hear where people are getting stuck.

There are two things I think should be mentioned in this guide, that are glossed over, and I think they would help clarify the situation.

`flux create ...` - whenever you run a command such as this, you are creating resources imperatively in the cluster. Instead, you can pass `--export` as an extra option to `flux create` subcommands, and the resource will be emitted on stdout instead, which you can capture and commit to a file somewhere that Flux can pick it up in your bootstrapped Flux repo structure.

An alternative to creating the resource via GitOps this way is to create it in the cluster, then use `flux export ...` to retrieve the definition, again, to commit back into the git repo and reify the cluster state with the repo state. This is the inversion of the traditional GitOps pattern, where resources are created in the repo first, vetted as a PR, then applied to the cluster when the branch merges.

"Flux for Helm Users" glosses over these things on purpose, because it's about teaching users how to use Helm with Flux, not about teaching the core concepts of Flux, but this is helpful feedback about the documentation; we can't predict the order that users will read our docs, and we can certainly try to do more to reinforce these concepts throughout the rest of the documentation for Flux.

Also, as far as I can tell you did not miss anything – the `traefik` namespace hasn't been created explicitly in the guide. It has to be created somewhere. This should at least get a mention. If you create it via GitOps, you may want to create it in the root Kustomization or in another Kustomization. Which you decide to do must take into account other considerations, like whether your cluster is set up for multi-tenancy or whether the Traefik (or other chart) is properly considered a tenant, or considered infrastructure. These are more considerations that seem likely to be placed outside of the scope of "Flux for Helm Users" – but surely they are addressed somewhere, and it might be good to mention that with a link to those docs.

You can also create the namespace imperatively, (but that will not make it reproducible via GitOps for disaster recovery.)

I don't know if we should create an issue or PR from this discussion to follow up, but hopefully this answers your questions for now!

**Reply by kingdonb** [reply]
I guess I should mention also; bootstrapping Flux with a PAT does not save the PAT, or give Flux any write access to the git repo after the bootstrap process is complete (unless you pass in `--read-write-key` at bootstrap time), nor will Flux ever try to write back to your repo unless you explicitly build out the structures that cause it to push updates (the `ImageUpdateAutomation` resource does this.)

The `flux create` commands will all create resources exclusively in your cluster, and not in the git repo. Yes, I agree this is backwards too, but not all people will use Flux in the same way, nor will all intend to follow GitOps principles "to the letter."

**Reply by RytisLT** [reply]
Cool, thanks for clearing things up. ~So the workflow is to use `flux create` with `--export` parameter, save output in repo and wait for reconciliation. If I want to "follow GitOps to the letter" then I would never use `flux create` without `--export` right? Because it sort of bypasses the git part. The wording in `Flux for Helm Users` made it sound like `--export` is an optional thing or maybe it's just me..~ You actually explained everything in post above, I've read the reply first, doh...
